### PR TITLE
7801 effect on discontinuous selection

### LINKED ISSUE #7801: Implement applying an effect to the selected clips
### Description

Implement the application of an effect to the selected clips, including the selection of non-consecutive clips.

QA: also test the ability to run a preview on a clip selected by double-clicking.

### PR DESCRIPTION
**Pull request by saintmatthieu**
Resolves: #7801

(Attempt 2 of https://github.com/audacity/audacity/pull/7835)

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/effects_base/internal/effectexecutionscenario.cpp [hunk 1]
```diff
@@ -302,6 +310,52 @@ muse::Ret EffectExecutionScenario::doPerformEffect(au3::Au3Project& project, con
...
+    // We are going to set the time and track selection to one clip at a time and apply the effect.
+
+    // Make a copy of the selection state and restore it when leaving this scope.
+    const trackedit::ClipKeyList clipsToProcess = selectionController()->selectedClips();
+    const trackedit::TrackIdList tracksToProcess = selectionController()->selectedTracks();
+
+    // No need to notify the UI about these changes, as they are temporary.
+    constexpr bool complete = false;
+
+    Finally restoreTrackSelection([&] {
```
Comment (Eism) [inline src/effects/effects_base/internal/effectexecutionscenario.cpp:326]: please use our Defer class: defer.h

File: src/effects/effects_base/internal/effectexecutionscenario.cpp [hunk 2]
```diff
@@ -82,22 +83,23 @@ muse::Ret EffectExecutionScenario::doPerformEffect(au3::Au3Project& project, con
...
+        }
+
+        if (selectionController()->hasSelectedClips()) {
+            // If multiple clips are selected, we have checked that the effect supports it, in which case these global time boundaries shouldn't be relevant.
+            // If this is just a single-clip selection, though, that will just be start and end times of the selected clip.
             t0 = selectionController()->selectedClipStartTime();
             t1 = selectionController()->selectedClipEndTime();
         } else {
             t0 = selectionController()->dataSelectedStartTime();
-            t1 = selectionController()->dataSelectedEndTime();
```
Comment (Eism) [inline src/effects/effects_base/internal/effectexecutionscenario.cpp:96]: could you please explain why you removed this?

Comment (saintmatthieu) [reply src/effects/effects_base/internal/effectexecutionscenario.cpp:96]: Mistake - good catch.

**Review by Eism — APPROVED** [review]

File: src/effects/effects_base/internal/effectexecutionscenario.cpp [hunk 3]
```diff
@@ -302,6 +312,52 @@ muse::Ret EffectExecutionScenario::doPerformEffect(au3::Au3Project& project, con
...
+    const trackedit::ClipKeyList clipsToProcess = selectionController()->selectedClips();
+    const trackedit::TrackIdList tracksToProcess = selectionController()->selectedTracks();
+
+    // No need to notify the UI about these changes, as they are temporary.
+    constexpr bool complete = false;
+
+    Defer restoreTrackSelection([&] {
+        selectionController()->setSelectedClips(clipsToProcess, complete);
+        selectionController()->setSelectedTracks(tracksToProcess, complete);
+    });
```
Comment (Eism) [inline src/effects/effects_base/internal/effectexecutionscenario.cpp:331]:
```suggestion
    DEFER {
        selectionController()->setSelectedClips(clipsToProcess, complete);
        selectionController()->setSelectedTracks(tracksToProcess, complete);
    };
```

Comment (saintmatthieu) [reply src/effects/effects_base/internal/effectexecutionscenario.cpp:331]: Why is this better ?

**Pull request merged**